ehrapy/api/preprocessing/_data_imputation.py

```diff
@@ -3,51 +3,236 @@
+        ) from None
+
+    return adata
+
+
+def _most_frequent(adata: AnnData, var_names: list[str] | None) -> AnnData:
+    imputer = SimpleImputer(strategy="most_frequent")
+    # impute a subset of columns
+    if isinstance(var_names, list):
+        column_indices = get_column_indices(adata, var_names)
+        adata.X[::, column_indices] = imputer.fit_transform(adata.X[::, column_indices])
+    # impute all columns if None passed
+    else:
+        adata.X = imputer.fit_transform(adata.X)
+
+    return adata
+
+
+# ===================== KNN Imputation =========================
+
+
+def _knn(adata: AnnData, var_names: list[str] | None) -> AnnData:
+    # numerical only data needs no encoding since KNN Imputation can be applied directly
+    if np.issubdtype(adata.X.dtype, np.number):
+        _knn_impute(adata, var_names)
+    else:
+        # ordinal encoding is used since non-numerical data can not be imputed using KNN Imputation
+        enc = OrdinalEncoder()
+        adata.X = enc.fit_transform(adata.X)
+        # impute the data using KNN imputation
+        _knn_impute(adata, var_names)
+        # decode ordinal encoding to obtain imputed original data
+        adata.X = enc.inverse_transform(adata.X)
+
+    return adata
+
+
+def _knn_impute(adata: AnnData, var_names: list[str] | None) -> None:
+    """Utility function to impute data using KNN-Imputer"""
+    imputer = KNNImputer(n_neighbors=1)
+
+    if isinstance(var_names, list):
+        column_indices = get_column_indices(adata, var_names)
+        adata.X[::, column_indices] = imputer.fit_transform(adata.X[::, column_indices])
+    # impute all columns if None passed
+    else:
+        adata.X = imputer.fit_transform(adata.X)
+
+
+def _miss_forest(adata: AnnData, var_names: dict[str, list[str]] | None) -> AnnData:
+    # var names got passed for faster indices lookup
+    if var_names:
+        # ensure both keys got passed together
+        try:
+            non_num_vars = var_names["non_numerical"]
+            num_vars = var_names["numerical"]
+        except KeyError:
+            raise MissForestKeyError(
+                "One or both of your keys provided for var_names are unknown. Only "
+                "numerical and non_numerical are available!"
+            )
+        # get the indices from the var names
+        non_num_indices = get_column_indices(adata, non_num_vars)
+        num_indices = get_column_indices(adata, num_vars)
 
-    # scenario 3: Replace all missing values in a subset of columns with a specified value per column
-    # TODO
+    # infer non numerical and numerical indices automatically
+    else:
+        non_num_indices_set = _get_non_numerical_column_indices(adata.X)
+        num_indices = [ind for ind in range(adata.X.shape[1]) if ind not in non_num_indices_set]
+        non_num_indices = list(non_num_indices_set)
+
+    imp_num = IterativeImputer(estimator=ExtraTreesRegressor(), initial_strategy="mean", max_iter=10, random_state=0)
```

Comment:
Should not parametrize the initial strategy here, since (from what I've read) its the common way for MissForest to use `mean`. 
We can parametrize `max_iter` of course, but one must be aware that, the larger this one got (and the more numerical columns `X` has, that will be imputed) this will result in an extrem long runtime.

Reply:
I would parameterize all of them with our defaults? Think that's reasonable.